File: pcie_host/tutorials/027_run_mla_only_int8/run_mla_only_int8.cpp
// Quantize on the host, run only the MLA over PCIe, and dequantize the INT8 results.
//
// Usage:
//   tutorial_027_run_mla_only_int8 --model yolo26n-det-int8-b1.tar.gz [--card 0]

#include <simaai/neat/pcie/Model.h>

#include <opencv2/imgcodecs.hpp>
#include <opencv2/imgproc.hpp>

#include <algorithm>
#include <cmath>
#include <cstdint>
#include <cstdlib>
#include <filesystem>
#include <iomanip>
#include <iostream>
#include <stdexcept>
#include <string>
#include <utility>
#include <vector>

namespace pcie = simaai::neat::pcie;

namespace {

constexpr int kBuildTimeoutMs = 180000;
constexpr int kRunTimeoutMs = 30000;
constexpr double kMaxErrorScales = 0.05; // one twentieth of a quantization step
constexpr char kImagePath[] = "share/sima-pcie-host/tutorials/assets/street-scene.png";

struct Args {
  std::string model;
  int card_id = 0;
};

std::string require_value(int argc, char** argv, int& index, const char* option) {
  if (index + 1 >= argc) {
    throw std::runtime_error(std::string("missing value for ") + option);
  }
  return argv[++index];
}

Args parse_args(int argc, char** argv) {
  Args args;
  for (int index = 1; index < argc; ++index) {
    const std::string arg = argv[index];
    if (arg == "--model") {
      args.model = require_value(argc, argv, index, "--model");
    } else if (arg == "--card") {
      args.card_id = std::stoi(require_value(argc, argv, index, "--card"));
    } else if (arg == "-h" || arg == "--help") {
      std::cout << "Usage: " << argv[0] << " --model yolo26n-det-int8-b1.tar.gz [--card 0]\n";
      std::exit(0);
    } else {
      throw std::runtime_error("unknown argument: " + arg);
    }
  }
  if (args.model.empty()) {
    throw std::runtime_error("--model is required: yolo26n-det-int8-b1.tar.gz from the Model Zoo");
  }
  return args;
}

std::string shape_string(const std::vector<std::int64_t>& shape) {
  std::string text = "[";
  for (std::size_t index = 0; index < shape.size(); ++index) {
    text += (index == 0 ? "" : ", ") + std::to_string(shape[index]);
  }
  return text + "]";
}

std::size_t element_count(const std::vector<std::int64_t>& shape) {
  std::size_t count = 1;
  for (const auto dim : shape) {
    count *= static_cast<std::size_t>(dim);
  }
  return count;
}

// The MLA-only contract quantizes every tensor with one scale and one zero point:
//   x = (q - zero_point) * scale
//   q = clamp(round(x / scale) + zero_point, -128, 127)
const pcie::QuantParams& require_quant(const pcie::TensorInfo& info) {
  if (!info.quant.has_value()) {
    throw std::runtime_error("tensor '" + info.name + "' publishes no quantization parameters");
  }
  return *info.quant;
}

std::int8_t quantize(const float value, const pcie::QuantParams& quant) {
  const float code = std::nearbyint(value / quant.scale) + static_cast<float>(quant.zero_point);
  return static_cast<std::int8_t>(std::clamp(code, -128.0F, 127.0F));
}

float dequantize(const std::int8_t code, const pcie::QuantParams& quant) {
  return static_cast<float>(static_cast<std::int32_t>(code) - quant.zero_point) * quant.scale;
}

const std::int8_t* int8_data(const pcie::Tensor& tensor) {
  return reinterpret_cast<const std::int8_t*>(static_cast<const std::uint8_t*>(tensor.data) +
                                              tensor.byte_offset);
}

// Preprocessing of the reference model: one RGB HWC image with pixels in [0, 1]. Resize, BGR to
// RGB, divide by 255, then quantize with the ingress parameters. Another model needs its own
// recipe.
std::vector<std::int8_t> quantize_image(const cv::Mat& bgr, const pcie::TensorInfo& input) {
  if (input.shape.size() != 3 || input.shape[2] != 3) {
    throw std::runtime_error("the reference model takes a three-channel HWC input, got " +
                             shape_string(input.shape));
  }
  const pcie::QuantParams& quant = require_quant(input);
  cv::Mat rgb;
  cv::resize(bgr, rgb,
             cv::Size(static_cast<int>(input.shape[1]), static_cast<int>(input.shape[0])));
  cv::cvtColor(rgb, rgb, cv::COLOR_BGR2RGB);
  std::vector<std::int8_t> codes(rgb.total() * rgb.channels());
  for (std::size_t index = 0; index < codes.size(); ++index) {
    codes[index] = quantize(static_cast<float>(rgb.data[index] / 255.0), quant);
  }
  return codes;
}

} // namespace

int main(int argc, char** argv) {
  try {
    const Args args = parse_args(argc, argv);
    if (!std::filesystem::is_regular_file(args.model)) {
      throw std::runtime_error("model does not exist: " + args.model);
    }
    const cv::Mat image = cv::imread(kImagePath, cv::IMREAD_COLOR);
    if (image.empty()) {
      throw std::runtime_error(std::string("OpenCV could not decode: ") + kImagePath);
    }
    pcie::ConnectionOptions connection;
    connection.card_id = args.card_id;

    // CORE LOGIC
    // STEP inspect-contract
    pcie::ModelOptions options;
    options.mla_only = true;
    pcie::Model model(args.model, options, connection);
    const pcie::ModelInfo info = model.info();
    if (info.inputs.size() != 1) {
      throw std::runtime_error("the reference model has exactly one input");
    }
    std::cout << "MLA-only contract:\n";
    for (const auto& input : info.inputs) {
      const pcie::QuantParams& quant = require_quant(input);
      std::cout << "  input " << input.name << " " << input.dtype << " "
                << shape_string(input.shape) << " scale=" << quant.scale
                << " zero_point=" << quant.zero_point << '\n';
    }
    for (const auto& output : info.outputs) {
      const pcie::QuantParams& quant = require_quant(output);
      std::cout << "  output " << output.name << " " << output.dtype << " "
                << shape_string(output.shape) << " scale=" << quant.scale
                << " zero_point=" << quant.zero_point << '\n';
    }
    // END STEP

    // STEP quantize-on-host
    const pcie::TensorInfo& ingress = info.inputs.front();
    std::vector<std::int8_t> codes = quantize_image(image, ingress);
    std::vector<float> fp32_input(codes.size());
    for (std::size_t index = 0; index < codes.size(); ++index) {
      fp32_input[index] = dequantize(codes[index], *ingress.quant);
    }
    pcie::TensorList int8_inputs;
    int8_inputs.push_back(pcie::Tensor::from_vector(std::move(codes), ingress.shape, ingress.name));
    // END STEP

    // STEP run-int8
    model.build(kBuildTimeoutMs);
    const pcie::TensorList outputs = model.run(int8_inputs, kRunTimeoutMs);
    model.close();
    // END STEP

    // STEP dequantize-and-compare
    pcie::Model reference(args.model, {}, connection);
    reference.build(kBuildTimeoutMs);
    pcie::TensorList fp32_tensors;
    fp32_tensors.push_back(
        pcie::Tensor::from_vector(std::move(fp32_input), ingress.shape, ingress.name));
    const pcie::TensorList reference_outputs = reference.run(fp32_tensors, kRunTimeoutMs);
    reference.close();

    std::cout << "Dequantized MLA-only outputs vs the default route (error in scale units):\n";
    for (std::size_t index = 0; index < outputs.size(); ++index) {
      const auto& spec = info.outputs[index];
      const pcie::QuantParams& quant = *spec.quant;
      const std::int8_t* codes = int8_data(outputs[index]);
      const auto& card = reference_outputs[index];
      if (card.route.name != spec.name || card.dtype != pcie::TensorDType::Float32) {
        throw std::runtime_error("default route output '" + spec.name + "' is not FP32");
      }
      const auto* card_values = reinterpret_cast<const float*>(
          static_cast<const std::uint8_t*>(card.data) + card.byte_offset);
      double max_error = 0.0;
      for (std::size_t element = 0; element < element_count(spec.shape); ++element) {
        const double host = dequantize(codes[element], quant);
        max_error = std::max(max_error, std::fabs(host - card_values[element]) / quant.scale);
      }
      std::cout << "  " << spec.name << " " << shape_string(spec.shape) << " max_err=" << std::fixed
                << std::setprecision(4) << max_error << '\n';
      if (max_error > kMaxErrorScales) {
        throw std::runtime_error("output '" + spec.name + "' deviates from the default route");
      }
    }
    // END STEP
    // END CORE LOGIC

    std::cout << "[OK] 027_run_mla_only_int8\n";
    return 0;
  } catch (const std::exception& error) {
    std::cerr << "[FAIL] " << error.what() << '\n';
    return 1;
  }
}

File: pcie_host/tutorials/027_run_mla_only_int8/run_mla_only_int8.py
#!/usr/bin/env python3
"""Quantize on the host, run only the MLA over PCIe, and dequantize the INT8 results.

Usage:
  python3 run_mla_only_int8.py --model yolo26n-det-int8-b1.tar.gz [--card 0]
"""

from __future__ import annotations

import argparse
from pathlib import Path

import cv2
import numpy as np

import pyneatpcie as pcie


BUILD_TIMEOUT_MS = 180_000
RUN_TIMEOUT_MS = 30_000
MAX_ERROR_SCALES = 0.05  # one twentieth of a quantization step
IMAGE_PATH = Path("share/sima-pcie-host/tutorials/assets/street-scene.png")


def parse_args() -> argparse.Namespace:
    parser = argparse.ArgumentParser(description=__doc__)
    parser.add_argument(
        "--model", required=True, help="yolo26n-det-int8-b1.tar.gz from the Model Zoo"
    )
    parser.add_argument("--card", type=int, default=0)
    return parser.parse_args()


def require_quant(spec: pcie.TensorInfo) -> tuple[float, int]:
    """Per-tensor parameters of the MLA-only contract.

    x = (q - zero_point) * scale
    q = clip(round(x / scale) + zero_point, -128, 127)
    """
    if spec.quant is None:
        raise RuntimeError(f"tensor '{spec.name}' publishes no quantization parameters")
    return spec.quant.scale, spec.quant.zero_point


def quantize(values: np.ndarray, scale: float, zero_point: int) -> np.ndarray:
    codes = np.rint(values.astype(np.float32) / np.float32(scale)) + zero_point
    return np.clip(codes, -128, 127).astype(np.int8)


def dequantize(codes: np.ndarray, scale: float, zero_point: int) -> np.ndarray:
    return (codes.astype(np.int32) - zero_point).astype(np.float32) * np.float32(scale)


def quantize_image(bgr: np.ndarray, spec: pcie.TensorInfo) -> np.ndarray:
    """Preprocessing of the reference model: one RGB HWC image with pixels in [0, 1].

    Resize, BGR to RGB, divide by 255. Another model needs its own recipe.
    """
    if len(spec.shape) != 3 or spec.shape[2] != 3:
        raise RuntimeError(f"the reference model takes a three-channel HWC input, got {spec.shape}")
    scale, zero_point = require_quant(spec)
    height, width, _ = spec.shape
    rgb = cv2.cvtColor(cv2.resize(bgr, (width, height)), cv2.COLOR_BGR2RGB)
    return quantize(rgb.astype(np.float64) / 255.0, scale, zero_point)


def main() -> None:
    args = parse_args()
    model_path = Path(args.model)
    if not model_path.is_file():
        raise FileNotFoundError(f"model does not exist: {model_path}")
    image = cv2.imread(str(IMAGE_PATH), cv2.IMREAD_COLOR)
    if image is None:
        raise RuntimeError(f"OpenCV could not decode: {IMAGE_PATH}")
    connection = pcie.ConnectionOptions(card_id=args.card)

    # CORE LOGIC
    # STEP inspect-contract
    options = pcie.ModelOptions()
    options.mla_only = True
    with pcie.Model(str(model_path), options, connection) as model:
        info = model.info()
        if len(info.inputs) != 1:
            raise RuntimeError("the reference model has exactly one input")
        print("MLA-only contract:")
        for spec in info.inputs:
            scale, zero_point = require_quant(spec)
            print(
                f"  input {spec.name} {spec.dtype} {spec.shape} "
                f"scale={scale} zero_point={zero_point}"
            )
        for spec in info.outputs:
            scale, zero_point = require_quant(spec)
            print(
                f"  output {spec.name} {spec.dtype} {spec.shape} "
                f"scale={scale} zero_point={zero_point}"
            )
        # END STEP

        # STEP quantize-on-host
        ingress = info.inputs[0]
        codes = quantize_image(image, ingress)
        scale, zero_point = require_quant(ingress)
        fp32_input = dequantize(codes, scale, zero_point)
        int8_inputs = [pcie.Tensor.from_numpy(codes, copy=True, route_name=ingress.name)]
        # END STEP

        # STEP run-int8
        model.build(BUILD_TIMEOUT_MS)
        outputs = model.run(int8_inputs, RUN_TIMEOUT_MS)
    codes_by_head = {spec.name: output.to_numpy() for output, spec in zip(outputs, info.outputs)}
    # END STEP

    # STEP dequantize-and-compare
    with pcie.Model(str(model_path), connection=connection) as reference:
        reference.build(BUILD_TIMEOUT_MS)
        fp32_tensor = pcie.Tensor.from_numpy(fp32_input, copy=True, route_name=ingress.name)
        reference_outputs = reference.run([fp32_tensor], RUN_TIMEOUT_MS)

    print("Dequantized MLA-only outputs vs the default route (error in scale units):")
    for card, spec in zip(reference_outputs, info.outputs):
        if card.route.name != spec.name:
            raise RuntimeError(f"default route output '{spec.name}' is missing")
        scale, zero_point = require_quant(spec)
        host = dequantize(codes_by_head[spec.name], scale, zero_point)
        max_error = float(np.max(np.abs(host - card.to_numpy())) / scale)
        print(f"  {spec.name} {spec.shape} max_err={max_error:.4f}")
        if max_error > MAX_ERROR_SCALES:
            raise RuntimeError(f"output '{spec.name}' deviates from the default route")
    # END STEP
    # END CORE LOGIC

    print("[OK] 027_run_mla_only_int8")


if __name__ == "__main__":
    main()
